src/download.ts

```diff
@@ -72,6 +72,15 @@ export class Download {
     return fileName;
   }
 
+  async saveAs(path: string) {
+    const internalPath = await this.path();
```

Comment:
You are racing the microtasks that are chained to `this.path()` in `saveAs` and `delete`. As a result, copy might happen after concurrently / after the unlink. I would rather store the request and await it in `_reportFinished` prior to calling `this._finishedCallback`.

Reply:
I see the race and agree it needs to change, but I'm not completely following the requested change.

What happens if you call `saveAs` multiple times?

```
await Promise.all([dl.saveAs('foo'), dl.saveAs('bar')]) 
```

What happens if you call `saveAs` after `_reportFinished` is already run?

Reply:
@pavelfeldman I made an attempt at it here: https://github.com/microsoft/playwright/pull/2872/commits/afe361ccef754ade7e19b0c50314310a2b04686e. I feel like I have one extra level of wrapping than needed. Also, is it ok that `_reportFinished` is now marked as `async`? The caller does not wait for it to resolve, but I think that's fine.